scripts/core/editor3/components/MultipleHighlights.jsx

```diff
@@ -3,260 +3,67 @@
-    }
-
-    styleNameBelongsToHighlight(styleName) {
-        var delimiterIndex = styleName.lastIndexOf('-');
-
-        if (delimiterIndex === -1) {
-            return false;
-        }
-
-        return Object.keys(this.props.availableHighlights).includes(styleName.slice(0, delimiterIndex));
-    }
-
-    getHighlightTypeFromStyleName(styleName) {
-        var delimiterIndex = styleName.lastIndexOf('-');
-
-        if (delimiterIndex === -1) {
-            throw new Error('Style name does not contain a highlight type');
-        }
-
-        var highlightType = styleName.slice(0, delimiterIndex);
-
-        if (this.highlightTypeValid(highlightType) === false) {
-            throw new Error(`Invalid highlight type '${highlightType}'`);
-        }
-
-        return highlightType;
-    }
-
-    getHighlightsCount(highlightType, editorState) {
-        const {newHighlightsState} = this.getHighlights(editorState);
-
-        if (highlightType !== undefined) {
-            if (this.highlightTypeValid(highlightType) === false) {
-                throw new Error(`Invalid highlight type '${highlightType}'`);
-            }
-            return newHighlightsState.lastHighlightIds[highlightType];
-        } else {
-            // count highlights of all types
-            return Object.keys(newHighlightsState.lastHighlightIds)
-                .reduce((count, key) => count + newHighlightsState.lastHighlightIds[key], 0);
-        }
-    }
-
-    getHighlightData(styleName, editorState) {
-        const {newHighlightsState} = this.getHighlights(editorState);
-
-        return newHighlightsState.highlightsData[styleName];
-    }
-
-    canAddtHighlight(highlightType, editorState) {
-        const {newEditorState} = this.getHighlights(editorState);
-
-        function characterHasAHighlightOfTheSameType(character) {
-            if (
-                character.getStyle()
-                    .some((styleName) =>
-                        this.styleNameBelongsToHighlight(styleName)
-                        && getHighlightType(styleName) === highlightType)
-            ) {
-                return true;
-            }
-
-            return false;
-        }
-
-        if (newEditorState.getSelection().isCollapsed()) {
-            return false;
-        }
-
-        // selection is expanded to include edges
-        // so you can't add a highlight right next to another
-        const selection = expandDraftSelection(
-            newEditorState.getSelection(),
-            newEditorState,
-            1,
-            1,
-            true
-        );
+    addHighlight(highlightType, highlightData) {
+        const editorState = Highlights.addHighlight(this.props.editorState, highlightType, highlightData);
 
-        return getDraftCharacterListForSelection(newEditorState, selection)
-            .some(characterHasAHighlightOfTheSameType.bind(this)) === false;
+        this.props.onHighlightChange(editorState);
```

Comment:
Are you calling `onChange` and returning the `editorState` at the same time? Shouldn't you do only one of the actions as it was previously? Wouldn't this result in multiple records in undo/redo stack when you are doing more processing with the `editorState` received from `addHighlight`?